docs/src/proposals/return-data.md

```diff
@@ -0,0 +1,139 @@
+    return (a, b, c);
+}
+```
+
+All the return values are eth abi encoded to a variable-length byte array.
+
+On ethereum errors can be returned too:
+
+```
+function withdraw() public {
+    require(msg.sender == owner, "Permission denied");
+}
+
+function failure() public {
+    revert("I afraid I can't do that dave");
+}
+```
+These errors help the developer debug any issue they are having, and can
+also be caught in a Solidity `try` .. `catch` block. Outside of a `try` .. `catch`
+block, any of these would cause the transaction or rpc to fail.
+
+## Existing solution
+
+The existing solution that Solang uses, writes the return data to the callee account data.
+The caller's account cannot be used, since the callee may not be the same BPF program, so
+it will not have permission to write to the callee's account data.
+
+Another solution would be to have a single return data account which is passed
+around through CPI. Again this does not work for CPI as the callee may not have
+permission to write to it.
+
+The problem with this solution is:
+
+- It does not work for RPC calls
+- It is very racey; a client has to submit the Tx and then retrieve the account
+  data. This is not atomic so the return data can be overwritten by another transaction.
+
+## Requirements for Solution
+
+It must work for:
+
+- RPC: An RPC should be able to return any number of values without writing to account data
+- Transaction: An transaction should be able to return any number of values without needing to write them account data
+- CPI: The callee must "set" return value, and the caller must be able to retrieve it.
+
+## Review of other chains
+
+### Ethereum (EVM)
+
+The `RETURN` opcode allows a contract to set a buffer as a returndata. This opcode takes a pointer to memory and a size. The `REVERT` opcode works similarly but signals that the call failed, and all account data changes must be reverted.
+
+For CPI, the caller can retrieve the returned data of the callee using the `RETURNDATASIZE` opcode which returns the length, and the `RETURNDATACOPY` opcode, which takes a memory destination pointer, offset into the returndata, and a length argument.
+
+Ethereum stores the returndata in blocks.
+
+### Parity Substrate
+
+The return data can be set using the `seal_return(u32 flags, u32 pointer, u32 size)` syscall.
+- Flags can be 1 for revert, 0 for success (nothing else defined)
+- Function does not return
+
+CPI: The `seal_call()` syscall takes pointer to buffer and pointer to buffer size where return data goes
+ - There is a 32KB limit for return data.
+
+Parity Substrate does not write the return data to blocks.
+
+## Rejected Solution
+
+The concept of ephemeral accounts has been proposed a solution for this. This would
+certainly work for the CPI case, but this would not work RPC or Transaction case.
+
+## Proposed Solution
+
+The callee can set the return data using a new system call `sol_set_return_data(u8 *buf, u64 length)`.
```

Comment:
Maybe this syscall could include "success" in its name to dissuade devs from mistakenly using it to return errors. Something like: `sol_return_success_with_data`

Reply:
I think the idea was that the non-success case could still return data, but it's obviously then only available by reading the transaction log over RPC

Reply:
> I think the idea was that the non-success case could still return data

The line below says:
> This function does not return and signals success for execution of the [instruction]

Does that not contradict what you're saying? Or are you referring to runtime verification failures?

Reply:
Indeed it does contradict what I said. @seanyoung, wdyt? I recall us discussing that it could be desirable for programs to have the ability to return data when they fail as well.

Reply:
> I think the idea was that the non-success case could still return data

Returning an error/non-success via this return buffer is a bad idea.
- Success/non-success becomes more complicated; program instruction can return success return code and an error in the return data
- On non-success the account data state changes should be reverted. We don't want to to this either on the VM side or program instruction side
- Only transaction that succeed should be valid, we don't intend on supporting recovery from errors (i.e. try-catch)
- Non-success cases can surface errors via program log.

Reply:
Any reason this syscall shouldn't be separated from success/fail and exit or not?  We could instead provide a syscall that allows the program to set some amount of "information data" which the program can set at any time but only the last data set is retained.  Effectively a single "return data" blob per tx.  This data could be stored next to the log data and available via rpc or the calling program and replaced within a tx every time an instruction processor calls the `set` syscall.  This would allow programs to return richer data on failure via rpc or return data to calling programs and rpc on success.

Reply:
That sounds like a great idea. I'll amend the proposal.

Reply:
BTW it would be great if we had a way of separating log data from return data. Could we have a special return data entry in the stable log? Ideally we could also have an event type entry too for solidity events.

Reply:
> Could we have a special return data entry in the stable log? Ideally we could also have an event type entry too for solidity events.

Yeah this is what I was thinking we'd do here